Downsampling unmapped text fields

salvatore-campagna opened the pull request:
When a field is unmapped usually dynamic mapping maps it
using a multi field which has the original field name as a text
field and a keyword sub-field. At downsampling time we skip
text fields and we only index the corresponding keyword field
in the target index. As a result, when indexing data into the
target index we need to use the name of the parent (text)
field instead of the (keyword) sub-field in order for indexing
to succeed.

Here we introduce a specific `MultiFieldValueFetcher` which
uses the name of the parent field as the field name. It derives
the name of the parent field by stripping away the name of
the sub-field (whatever appears after the last '.' in the name).

Here I also add a couple of yaml tests which test the behaviour
of text fields used in TSDB indices with synthetic source enabled.

Resolves #94346

salvatore-campagna committed:
fix: downsampling unmapped text fields

When a field is unmapped usually dynamic mapping maps it using
a multi field which has the original field name as a text field
and a keyword sub-field. At downsampling time we skip text fields
and we only index the corresponding keyword field in the target index.
As a result, when indexing data into the target index we need to
use the name of the parent (text) field instead of the (keyword)
sub-field in order for indexing to succeed.

Here we introduce a specific MultiFieldValueFetcher which uses the
name of the parent field as the field name. It derives the name
of the parent field by stripping away the name of the sub-field
(whatever appers after the last '.' in the name).

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
Update docs/changelog/94387.yaml

elasticsearchmachine commented:
Hi @salvatore-campagna, I've created a changelog YAML for you.

elasticsearchmachine commented:
Hi @salvatore-campagna, I've updated the changelog YAML for you.

salvatore-campagna committed:
Update docs/changelog/94387.yaml

salvatore-campagna committed:
fix: check search response results

salvatore-campagna commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/10_basic.yml:
```diff
@@ -961,3 +1027,27 @@ setup:
+        body:
+          sort: [ "_tsid", "@timestamp" ]
+
+  - length: { hits.hits: 4 }
```
Adding some `match` here.

martijnvg commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/FieldValueFetcher.java:
```diff
@@ -25,9 +25,9 @@
-    private final MappedFieldType fieldType;
-    private final IndexFieldData<?> fieldData;
-    private final AbstractDownsampleFieldProducer rollupFieldProducer;
+    protected final MappedFieldType fieldType;
```
can the visibility of these fields remain private?

martijnvg commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/90_unsupported_operations.yml:
```diff
@@ -232,3 +232,91 @@ aggregate on _id:
+---
+synthetic source text field:
+  - skip:
+      version: " - 8.6.99"
```
Synthetic source was made the default source mode in 8.7.0 for tsdb.

martijnvg commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/90_unsupported_operations.yml:
```diff
@@ -232,3 +232,91 @@ aggregate on _id:
+      reason: "synthetic source introduced in 8.7.0"
+
+  - do:
+      catch: /field \[k8s.pod.labels\] of type \[flattened\] doesn't support synthetic source/
```
we should be fixing that soon :)

csoulios replied:
Is synthetic source the only reason that flattened fields are not supportes in time-series indices?

I think it is hard to support flattened fields as dimensions or metrics. They can only exist as labels, right?

martijnvg replied:
I think sub fields of flattened field should be able to be dimensions. This doesn't work now. But I think we can support that too? I don't think we can support subfields of flattened fields to be metrics.

martijnvg reviewed (approved):
LGTM

csoulios commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/90_unsupported_operations.yml:
```diff
@@ -232,3 +232,91 @@ aggregate on _id:
+      reason: "synthetic source introduced in 8.7.0"
+
+  - do:
+      catch: /field \[k8s.pod.agent.name\] of type \[text\] doesn't support synthetic source unless it is stored or has a sub-field of type \[keyword\] with doc values or stored and without a normalizer/
+      indices.create:
```
I believe that not supporting `text` fields is a bigger problem than not supporting `flattened fields`. We should be able to support `text` fields as labels

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/MultiFieldValueFetcher.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public String name() {
+        return fieldType.name().substring(0, fieldType.name().lastIndexOf('.'));
```
Can we cache the result of this function in an instance variable? I am afraid this may be called way too often to compute it every time.

salvatore-campagna replied:
That makes sense.

csoulios reviewed (approved):
LGTM

I only left an comment about not computing the name of the field at every call.

salvatore-campagna committed:
fix: cache the field name

We can't set the name if the MultiFieldValueFetcher constructor
because it is used in the parent constructor. So we get rid of
MultiFieldValueFetcher and just cache the name, setting it propertly
for flattened fields such that the parent field name is used.

salvatore-campagna committed:
fix: cache the field name

salvatore-campagna committed:
docs: include a note about multi-valued fields

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/part-1 please

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.7](https://github.com/elastic/elasticsearch/pull/94514)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/94514">](https://github.com/elastic/elasticsearch/pull/94514) |
